### PR TITLE
fix inconsistent dashboard action buttons hover effects

### PR DESCRIPTION
**Pull request by WiNloSt**
Epic #22826

Task: https://www.notion.so/metabase/Inconsistent-hover-effects-on-action-buttons-5e619a2f52e04765872e82f424260bdd

### CODE REVIEW

File: frontend/src/metabase/collections/components/CollectionHeader/CollectionHeader.styled.tsx [hunk 1]
```diff
@@ -20,3 +22,18 @@ export const HeaderActions = styled.div`
...
+  height: 2.5rem;
+
+  &:hover {
+    color: ${color("brand")};
+    background-color: ${color("bg-medium")};
+  }
+`;
+
+CollectionHeaderButton.defaultProps = {
+  onlyIcon: true,
```
Comment (WiNloSt) [inline frontend/src/metabase/collections/components/CollectionHeader/CollectionHeader.styled.tsx:37]: Hide the border.

File: frontend/src/metabase/core/components/BookmarkToggle/BookmarkToggle.tsx [hunk 2]
```diff
@@ -44,10 +38,9 @@ const BookmarkToggle = forwardRef(function BookmarkToggle(
 
   return (
     <Tooltip tooltip={isBookmarked ? t`Remove from bookmarks` : t`Bookmark`}>
-      <IconWrapper
-        {...props}
+      <BookmarkButton
```
Comment (WiNloSt) [inline frontend/src/metabase/core/components/BookmarkToggle/BookmarkToggle.tsx:41]: Change to button to allow selection via <kbd>Tab</kbd>

File: frontend/src/metabase/core/components/Button/Button.styled.tsx [hunk 3]
```diff
@@ -2,6 +2,7 @@ import styled from "@emotion/styled";
 import { css } from "@emotion/react";
 
 export const ButtonRoot = styled.button`
+  transition: all 200ms linear;
```
Comment (WiNloSt) [inline frontend/src/metabase/core/components/Button/Button.styled.tsx:5]: Button without `Button` class won't transition at all.

https://github.com/metabase/metabase/blob/0b050998995894c06875b03b6ecd0d90e721a263/frontend/src/metabase/css/components/buttons.css#L24

File: frontend/src/metabase/core/components/Link/Link.tsx [hunk 4]
```diff
@@ -19,14 +19,15 @@ const Link = ({
...
+  tabIndex,
   ...props
 }: LinkProps): JSX.Element => {
   const link = (
     <LinkRoot
       {...props}
       to={to}
       disabled={disabled}
-      tabIndex={disabled ? -1 : undefined}
+      tabIndex={tabIndex ?? disabled ? -1 : undefined}
```
Comment (WiNloSt) [inline frontend/src/metabase/core/components/Link/Link.tsx:30]: Allow overriding the `tabIndex`. This is used in https://github.com/metabase/metabase/pull/23833/files#diff-eeeb01acb404315c0c6c49673896cc649a1339bd90341aaced2e75eb9c9d4af0R19 to prevent <kbd>Tab</kbd> selection.

File: frontend/src/metabase/dashboard/components/DashboardActions.jsx [hunk 5]
```diff
@@ -54,15 +53,12 @@ export const getDashboardActions = (
...
-          <span>
-            <DashboardHeaderButton
-              disabled={!canManageSubscriptions}
-              onClick={onSharingClick}
-              data-metabase-event="Dashboard;Subscriptions"
-            >
-              <Icon size={18} name="subscription" />
-            </DashboardHeaderButton>
-          </span>
+          <DashboardHeaderButton
```
Comment (WiNloSt) [inline frontend/src/metabase/dashboard/components/DashboardActions.jsx:56]: Remove span and change the underlying implementation of `DashboardHeaderButton` to base off `Button` component which supports <kbd>Tab</kbd> selection

File: frontend/src/metabase/dashboard/components/DashboardHeader.styled.tsx [hunk 6]
```diff
@@ -149,13 +149,6 @@ export const HeaderButtonsContainer = styled.div<TypeForItemsThatRespondToNavBar
     padding-right: 1rem;
     border-top: 1px solid ${color("border")};
   }
-
-  ${Button.Root} {
-    padding: 0.5rem 0.75rem;
-    &:hover {
-      color: ${color("brand")};
-    }
-  }
```
Comment (WiNloSt) [inline frontend/src/metabase/dashboard/components/DashboardHeader.styled.tsx:158]: Moved to `frontend/src/metabase/dashboard/containers/DashboardHeader.styled.jsx` so we have style that target the same elements in the same file.

File: frontend/src/metabase/dashboard/components/DashboardHeader.tsx [hunk 7]
```diff
@@ -85,20 +85,14 @@ const DashboardHeader = ({
...
-              key={sectionIndex}
-              className="Header-buttonSection"
-              isNavBarOpen={isNavBarOpen}
-            >
-              {section}
-            </HeaderButtonSection>
-          )
-        );
-      }),
+    () => (
```
Comment (WiNloSt) [inline frontend/src/metabase/dashboard/components/DashboardHeader.tsx:88]: Previously, it was not obvious what we do here. The `headerButtons` looked like this.

```ts
[[button1, button2, button3]]
```

You could see that we only sent `buttons` in an array making it an array within another array
https://github.com/metabase/metabase/blob/a26cba2a44eae19b92cda37356f28fd9cf6f9af3/frontend/src/metabase/dashboard/containers/DashboardHeader.jsx#L370.

This is simplified now. https://github.com/metabase/metabase/pull/23833/files#diff-353d899c9d83acd25f94eddc7773d8377fbe420ed1bae75084691d0d292f2210R363

File: frontend/src/metabase/dashboard/containers/DashboardHeader.jsx [hunk 8]
```diff
@@ -209,12 +208,11 @@ class DashboardHeader extends Component {
       buttons.push(
         <Tooltip tooltip={addQuestionButtonHint}>
           <DashboardHeaderButton
+            icon="add"
```
Comment (WiNloSt) [inline frontend/src/metabase/dashboard/containers/DashboardHeader.jsx:211]: `DashboardHeaderButton` is not based off `Button`.

File: frontend/src/metabase/dashboard/containers/DashboardHeader.jsx [hunk 9]
```diff
@@ -285,16 +283,13 @@ class DashboardHeader extends Component {
     if (!isFullscreen && !isEditing && canEdit) {
       buttons.push(
         <Tooltip key="edit-dashboard" tooltip={t`Edit dashboard`}>
-          <a
-            data-metabase-event="Dashboard;Edit"
+          <DashboardHeaderButton
```
Comment (WiNloSt) [inline frontend/src/metabase/dashboard/containers/DashboardHeader.jsx:286]: Not sure why this is implemented using an anchor tag as it didn't have the `href` attribute.

File: frontend/src/metabase/dashboard/containers/DashboardHeader.jsx [hunk 10]
```diff
@@ -344,19 +339,17 @@ class DashboardHeader extends Component {
...
-            onlyIcon
-            borderless
-            isShowingDashboardInfoSidebar={isShowingDashboardInfoSidebar}
-            onClick={() =>
-              isShowingDashboardInfoSidebar
-                ? closeSidebar()
-                : setSidebar({ name: SIDEBAR_NAME.info })
-            }
-          />,
+          <Tooltip key="dashboard-info-button" tooltip={t`More info`}>
```
Comment (WiNloSt) [inline frontend/src/metabase/dashboard/containers/DashboardHeader.jsx:342]: Add a missing tooltip to match the button on questions.
#### Dashboard
image

#### Questions
image

File: frontend/src/metabase/dashboard/containers/DashboardHeader.jsx [hunk 11]
```diff
@@ -344,19 +339,17 @@ class DashboardHeader extends Component {
...
+            <DashboardHeaderButton
```
Comment (WiNloSt) [inline frontend/src/metabase/dashboard/containers/DashboardHeader.jsx:343]: Use `DashboardHeaderButton` the same as other buttons in the dashboard header.

File: frontend/src/metabase/dashboard/containers/DashboardHeader.styled.jsx [hunk 12]
```diff
@@ -34,7 +11,22 @@ export const DashboardHeaderActionDivider = styled.div`
   border-left: 1px solid ${color("border-dark")};
 `;
 
-export const DashboardHeaderInfoButton = styled(Button)`
-  color: ${props =>
-    props.isShowingDashboardInfoSidebar ? color("brand") : color("text-dark")};
+export const DashboardHeaderButton = styled(Button)`
```
Comment (WiNloSt) [inline frontend/src/metabase/dashboard/containers/DashboardHeader.styled.jsx:14]: Combine both `DashboardHeaderButton` and `DashboardHeaderInfoButton` into `DashboardHeaderButton` as they have similar behaviors.

File: frontend/src/metabase/nav/components/ProfileLink.jsx [hunk 13]
```diff
@@ -106,11 +105,7 @@ function ProfileLink({ user, adminItems, onLogout }) {
         items={generateOptionsForUser()}
         triggerIcon="gear"
         triggerProps={{
-          color: color("text-medium"),
-          hover: {
-            backgroundColor: color("brand"),
-            color: color("text-white"),
-          },
+          isInvertedColor: true,
```
Comment (WiNloSt) [inline frontend/src/metabase/nav/components/ProfileLink.jsx:108]: Combine multiple explicit props into a single variant so it's easier to implement the styled component.

File: frontend/src/metabase/core/components/Link/Link.tsx (hunk 4 above)
Comment (kulyk) [reply frontend/src/metabase/core/components/Link/Link.tsx:30]: This expression feels a bit confusing to me. My first thought was that `tabIndex ?? disabled` is the ternary condition. However, it's actually `tabIndex` OR `disabled ? -1 : undefined`, right? Should we maybe extract this block? Something like `useMemo` can make it a bit more straightforward I believe:

```js
const tabIndex = useMemo(() => {
  if (tabIndex) {
     return tabIndex;
  }
  return disabled ? -1 : undefined;
}, [ ... ])
```

File: frontend/src/metabase/components/Icon.tsx [hunk 14]
```diff
@@ -23,7 +23,7 @@ export const IconWrapper = styled.div<IconWrapperProps>`
   justify-content: center;
   width: 40px;
   height: 40px;
-  border-radius: 99px;
+  border-radius: 6px;
```
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/components/Icon.tsx:26]: I wonder if the change might cause visual regressions? Don't we expect some icons to be round?

File: frontend/src/metabase/components/EntityMenuTrigger.styled.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    return (
+      isInvertedColor &&
+      `color: ${color("text-light")};
+
+      &:hover {
+        color: ${color("text-white")};
+        background-color: ${color("brand")};
+      }`
+    );
+  }}
```
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/components/EntityMenuTrigger.styled.ts:28]: I'm not sure if we're going to have more components that are going to use `isInvertedColor`. Colors in there are hardcoded, so it's a bit limiting (although figuring out colors in a "smart way" feels too smart 😄). Should we maybe just keep things as they were when `ProfileLink` was passing styles directly? `EntityMenu` could use a refactoring and general cleanup, I believe we should end up with an API that'd allow styling stuff via `emotion`, e.g.

```jsx
const StyledEntityMenu = styled(EntityMenu)`
   ${EntityMenu.Trigger} {
      /* styles */
   }
`
```

No need in a refactoring right now for sure, just thinking about how it should probably work to match our approach in other components

File: frontend/src/metabase/collections/components/CollectionHeader/CollectionTimeline.tsx [hunk 16]
```diff
@@ -16,13 +16,11 @@ const CollectionTimeline = ({
...
-    <Tooltip tooltip={t`Events`}>
-      <Link to={url}>
-        <IconWrapper>
-          <Icon name="calendar" size={20} />
-        </IconWrapper>
-      </Link>
-    </Tooltip>
+    <Link to={url} tabIndex={-1}>
+      <Tooltip tooltip={t`Events`}>
+        <CollectionHeaderButton icon="calendar" iconSize={20} onlyIcon />
```
Comment (kulyk) [inline frontend/src/metabase/collections/components/CollectionHeader/CollectionTimeline.tsx:21]: Guess there is no need to pass `onlyIcon` prop here if it's already defined as a default one for `CollectionHeaderButton`?

Comment (kulyk) [reply frontend/src/metabase/collections/components/CollectionHeader/CollectionTimeline.tsx:21]: `iconSize` can also be a default prop. Or we can remove it from `CollectionHeaderButton's` `defaultProps` and pass it explicitly as it is now

**Review by kulyk — COMMENTED** [review]
Looking good! Left a few small comments. Also noticed that for questions, the "more" button's hover area is much bigger than for other actions (more, bookmark, etc.), take a look at the GIF:

CleanShot 2022-07-12 at 12 36 15

File: frontend/src/metabase/components/Icon.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase/components/Icon.tsx:26]: @mazameli Did I go overboard? I interpreted that from https://www.notion.so/metabase/Inconsistent-hover-effects-on-action-buttons-5e619a2f52e04765872e82f424260bdd we're gonna use rounded rectangles instead of circular buttons/icons.

File: frontend/src/metabase/components/EntityMenuTrigger.styled.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase/components/EntityMenuTrigger.styled.ts:28]: That's fair. I'll revert the change in `ProfileLink` to pass styles directly.

File: frontend/src/metabase/core/components/Link/Link.tsx (hunk 4 above)
Comment (WiNloSt) [reply frontend/src/metabase/core/components/Link/Link.tsx:30]: Yeah, you're right. It might be obvious to me now, but this could confuse other people. +1 for simplifying this.

File: frontend/src/metabase/collections/components/CollectionHeader/CollectionTimeline.tsx (hunk 16 above)
Comment (WiNloSt) [reply frontend/src/metabase/collections/components/CollectionHeader/CollectionTimeline.tsx:21]: Ahhh yeah, good catch!

File: frontend/test/metabase/scenarios/organization/timelines-collection.cy.spec.js [hunk 17]
```diff
@@ -91,7 +92,11 @@ describe("scenarios > organization > timelines > collection", () => {
       cy.findByText("Add an event").click();
 
       cy.findByLabelText("Event name").type("RC1");
-      cy.findByRole("button", { name: "calendar icon" }).click();
+
+      // New event modal
+      modal().within(() => {
+        cy.findByRole("button", { name: "calendar icon" }).click();
```
Comment (WiNloSt) [inline frontend/test/metabase/scenarios/organization/timelines-collection.cy.spec.js:98]: With changes within this PR, this command will find 2 calendar buttons.
1. Inside the new event modal
2. On the collection page (see the screenshot)
image

File: frontend/src/metabase/components/Icon.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (mazameli) [reply frontend/src/metabase/components/Icon.tsx:26]: I'll take a look.

**Review by kulyk — APPROVED** [review]
LGTM! Nice work

**Comment by mazameli** [comment]
Hover state on this button looks wrong:

image

**Comment by mazameli** [comment]
The height and width of these seem to be a bit all-over-the-place:

hovers-on-buttons

**Comment by WiNloSt** [comment]
@mazameli Are you on the latest commit 18922e970a1bc8e943e23f87aafcd875f9903561?

> Hover state on this button looks wrong:
> 
> image

> The height and width of these seem to be a bit all-over-the-place:
> 
> hovers-on-buttons


The screenshot & gif look like something I encountered during the development of this PR, but it shouldn't be like that anymore.
Here's what it should look like at the moment.
Jul-12-2022 23-27-09

By the way, I haven't touched anything button on the question header except the more actions button `...`. I noticed the buttons on the question header are smaller (16px) than the dashboard (18px) and collections (20px) one. I could try to make these consistent, but I was only trying to get the dashboard more consistent first (having the same hover style + tabbable).

I think we should try to address the collection, dashboard, and question buttons consistency separately, as this PR is already big enough. What do you think?

**Comment by mazameli** [comment]
> are you on the latest commit?

D'oh 🤦🏼  — I was not, sorry. Let me look again now.

**Comment by mazameli** [comment]
Here's what I'm seeing now. Still a couple inconsistencies:

inconsistencies

**Comment by WiNloSt** [comment]
@mazameli Yup, on questions. The only button that's touched is the more actions button `...`. Actually, I should have asked you right at the beginning, but I just assumed from the screenshots here https://www.notion.so/metabase/Inconsistent-hover-effects-on-action-buttons-5e619a2f52e04765872e82f424260bdd. Those are dashboard buttons which are first focused in this PR. 

And like I've explained in [the previous comment](https://github.com/metabase/metabase/pull/23833#issuecomment-1181992448), this PR is getting pretty large. I thought initially it's gonna be a relatively small PR. Do you think it's sensible to ship this PR now and create a few more follow-up PRs on making these buttons on collections, dashboads, and questions more consistent?

In summary, this PR
- Make dashboard icon buttons more consistent in style and make them all tabbable.
- Address minor icon buttons inconsistency on questions and collections, most prominently the More button is reused across all these 3 entities that's why it looks bigger on the questions header.

What's left. Let me know if this looks comprehensive.
- Make collections, dashboards, and question buttons more consistent in:
   - size
   - background color

File: frontend/src/metabase/timelines/common/components/ModalBody/ModalBody.tsx [hunk 18]
```diff
@@ -6,7 +6,7 @@ export interface ModalBodyProps {
 }
 
 const ModalBody = ({ children }: ModalBodyProps): JSX.Element => {
-  return <BodyRoot>{children}</BodyRoot>;
+  return <BodyRoot className="ModalContent">{children}</BodyRoot>;
```
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/src/metabase/timelines/common/components/ModalBody/ModalBody.tsx:9]: We use `cy.get(".Modal")` or (variations) for custom modals. I recall this class cannot be added here.

File: frontend/src/metabase/collections/components/CollectionHeader/CollectionTimeline.tsx [hunk 19]
```diff
@@ -16,13 +16,11 @@ const CollectionTimeline = ({
...
 
   return (
-    <Tooltip tooltip={t`Events`}>
-      <Link to={url}>
-        <IconWrapper>
-          <Icon name="calendar" size={20} />
-        </IconWrapper>
-      </Link>
-    </Tooltip>
+    <Link to={url} tabIndex={-1}>
```
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/src/metabase/collections/components/CollectionHeader/CollectionTimeline.tsx:19]: It's incorrect to have `button` inside `a`. More info here https://stackoverflow.com/questions/6393827/can-i-nest-a-button-element-inside-an-a-using-html5.

File: frontend/src/metabase/core/components/BookmarkToggle/BookmarkToggle.tsx [hunk 20]
```diff
@@ -7,24 +7,18 @@ import React, {
...
   onCreateBookmark: () => void;
   onDeleteBookmark: () => void;
 }
 
 const BookmarkToggle = forwardRef(function BookmarkToggle(
-  {
-    isBookmarked,
-    onCreateBookmark,
-    onDeleteBookmark,
-    ...props
```
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/src/metabase/core/components/BookmarkToggle/BookmarkToggle.tsx:25]: This is a standard pattern for base components that support ref forwarding, mostly to support passing `aria-` and `data-` attributes. Take a look at other components in `/core/components` - most of them follow this convention.

File: frontend/src/metabase/core/components/BookmarkToggle/BookmarkToggle.tsx [hunk 21]
```diff
@@ -44,10 +38,9 @@ const BookmarkToggle = forwardRef(function BookmarkToggle(
 
   return (
     <Tooltip tooltip={isBookmarked ? t`Remove from bookmarks` : t`Bookmark`}>
-      <IconWrapper
-        {...props}
+      <BookmarkButton
         ref={ref}
-        hover={{ color: isBookmarked ? color("brand") : color("text-dark") }}
+        hoverColor={isBookmarked ? color("brand") : color("text-dark")}
```
Comment (ranquild) [inline frontend/src/metabase/core/components/BookmarkToggle/BookmarkToggle.tsx:43]: Can we pass `isBookmarked` to the styled component instead? We try to use `color` in styled components only

File: frontend/src/metabase/core/components/Link/Link.tsx [hunk 22]
```diff
@@ -19,14 +24,23 @@ const Link = ({
   children,
   disabled,
   tooltip,
+  tabIndex,
   ...props
 }: LinkProps): JSX.Element => {
+  const computedTabIndex = useMemo(() => {
```
Comment (ranquild) [inline frontend/src/metabase/core/components/Link/Link.tsx:30]: `useMemo` won't give any performance gains here, and actually it would decrease performance. Could it be removed?

**Review by ranquild — CHANGES_REQUESTED** [review]
Looks good overall. Requesting changes mainly due to timeline changes with links and inner buttons. Could this change be reverted so we can address it separately? Thanks!

File: frontend/src/metabase/core/components/BookmarkToggle/BookmarkToggle.tsx (hunk 21 above)
Comment (WiNloSt) [reply frontend/src/metabase/core/components/BookmarkToggle/BookmarkToggle.tsx:43]: Fixed.

File: frontend/src/metabase/collections/components/CollectionHeader/CollectionTimeline.tsx [hunk 23]
```diff
@@ -16,11 +16,11 @@ const CollectionTimeline = ({
...
 
   return (
-    <Link to={url} tabIndex={-1}>
-      <Tooltip tooltip={t`Events`}>
-        <CollectionHeaderButton icon="calendar" />
-      </Tooltip>
-    </Link>
+    <Tooltip tooltip={t`Events`}>
+      <div>
+        <CollectionHeaderButton as={Link} to={url} icon="calendar" />
```
Comment (WiNloSt) [inline frontend/src/metabase/collections/components/CollectionHeader/CollectionTimeline.tsx:21]: This will only render a single `<a>` tag, so no more `<button />` inside an `<a />` tag.

File: frontend/src/metabase/core/components/Button/Button.tsx [hunk 24]
```diff
@@ -77,6 +85,7 @@ const BaseButton = forwardRef(function BaseButton(
 
   return (
     <ButtonRoot
+      as={as}
```
Comment (WiNloSt) [inline frontend/src/metabase/core/components/Button/Button.tsx:88]: Allow render button as another component e.g. `Link`

File: frontend/src/metabase/core/components/Link/Link.tsx [hunk 25]
```diff
@@ -27,20 +22,12 @@ const Link = ({
...
-    return disabled ? -1 : undefined;
-  }, [disabled, tabIndex]);
-
   const link = (
     <LinkRoot
       {...props}
       to={to}
       disabled={disabled}
-      tabIndex={computedTabIndex}
+      tabIndex={disabled ? -1 : undefined}
```
Comment (WiNloSt) [inline frontend/src/metabase/core/components/Link/Link.tsx:30]: revert change

**Review by mazameli — APPROVED** [review]
Nice!

**Review by ranquild — APPROVED** [review]
